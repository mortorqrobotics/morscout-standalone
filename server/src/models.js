import { model } from "mongoose";
import allianceSchema from "shared/schemas/Alliance";
import matchSchema from "shared/scheams/Match";
import regionalSchema from "shared/schemas/Regional";
import seasonSchema from "shared/scheams/Season";
import teamSchema from "shared/schemas/Team";
import userSchema from "shared/schemas/User";

export const Alliance = model("Alliance", allianceSchema);
export const Match = model("Match", matchSchema);
export const Regional = model("Regional", regionalSchema);
export const Season = model("Season", seasonSchema);
export const Team = model("Team", teamSchema);
export const User = model("User", userSchema);
