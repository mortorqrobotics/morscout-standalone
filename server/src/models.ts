import { model, Model } from "mongoose";
import allianceSchema, { IAlliance } from "Shared/schemas/Alliance";
import matchSchema, { IMatch } from "Shared/schemas/Match";
import regionalSchema, { IRegional } from "shared/schemas/Regional";
import seasonSchema, { ISeason } from "Shared/schemas/Season";
import teamSchema, { ITeam } from "Shared/schemas/Team";
import userSchema, { IUser } from "Shared/schemas/User";

export const Alliance: Model<IAlliance> = model("Alliance", allianceSchema);
export const Match: Model<IMatch> = model("Match", matchSchema);
export const Regional: Model<IRegional> = model("Regional", regionalSchema);
export const Season: Model<ISeason> = model("Season", seasonSchema);
export const Team: Model<ITeam> = model("Team", teamSchema);
export const User: Model<IUser> = model("User", userSchema);
