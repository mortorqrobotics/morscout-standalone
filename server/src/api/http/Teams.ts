import getTeams from "api/Teams/getTeams";
import getTeam from "api/Teams/getTeam";
import { Router, Request, Response } from "express";

export default (router: Router) => {
  router.get("/teams/:season", async (req: Request, res: Response) => {
    res.end(JSON.stringify(await getTeams(req.params.season)));
  });
  router.get("/team/:team", async (req: Request, res: Response) => {
    res.end(JSON.stringify(await getTeam(req.params.team)));
  });
};
