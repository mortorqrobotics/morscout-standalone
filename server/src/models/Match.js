import { Schema } from "mongoose";

const Match = new Schema({
  redAlliance: {
    type: Schema.Types.ObjectId,
    ref: "Alliance"
  },
  blueAlliance: {
    type: Schema.Types.ObjectId,
    ref: "Alliance"
  },
  winner: {
    type: String,
    enum: ["Red", "Blue"],
    default: null
  },
  startDate: {
    type: Date
  },
  endData: {
    type: Date,
    validate: date => this.startDate.time() < date.time()
  },
  Regional: {
    type: Schema.Types.ObjectId,
    ref: "Regional"
  }
});

export default Match;
