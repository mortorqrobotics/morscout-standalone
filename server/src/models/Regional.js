import { Schema } from "mongoose";
import { pointSchema, polygonSchema } from "./util/GeoJSON";
import { eventTypes } from "./util/TBA";

export default mongoose => {
  const RegionalSchema = new Schema({
    name: String,
    location: pointSchema,
    district: polygonSchema,
    url: {
      type: String,
      validate: url => encodeURI(url) === url
    },
    eventType: {
      type: Number,
      validate: type => Object.keys(eventTypes).includes(type)
    },
    startDate: {
      type: Date
    },
    endData: {
      type: Date,
      validate: date => this.startDate.time() < date.time()
    },
    season: {
      type: Schema.types.ObjectId,
      ref: "Season"
    }
  });

  return mongoose.model("Regional", RegionalSchema);
};
