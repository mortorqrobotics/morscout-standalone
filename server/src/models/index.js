import * as mongoose from "mongoose";

import MatchSchema from "./Match";
import RegionalSchema from "./Regional";
import TeamSchema from "./Team";
import UserSchema from "./User";

export const Match = mongoose.model("Match", MatchSchema);
export const Regional = mongoose.model("Regional", RegionalSchema);
export const Team = mongoose.model("Team", TeamSchema);
export const User = mongoose.model("User", UserSchema);
