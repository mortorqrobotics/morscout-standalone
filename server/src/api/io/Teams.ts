import { updated, loadTeams, loadTeam } from "shared/types/Teams";
import { Schema } from "mongoose";
import { Socket } from "socket.io";
import getTeams from "api/Teams/getTeams";
import getTeam from "api/Teams/getTeam";

export default (socket: Socket) => {
  socket.on(loadTeams, async () => {
    socket.emit("action", {
      type: updated,
      data: await getTeams(socket)
    });
  });
  socket.on(loadTeam, async (id: Schema.Types.ObjectId) => {
    socket.emit("action", {
      type: updated,
      data: {
        [id.toString()]: await getTeam(id)
      }
    });
  });
};
