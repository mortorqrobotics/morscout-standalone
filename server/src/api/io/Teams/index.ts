import { updated, loadTeams, loadTeam } from "Shared/types/Teams";
import { Schema } from "mongoose";
import { Socket } from "socket.io";
import getTeams from "./getTeams";
import getTeam from "./getTeam";

export default (socket: Socket) => {
  socket.on(loadTeams, async () => {
    socket.emit("action", {
      type: updated,
      data: await getTeams(socket)
    });
  });
  socket.on(loadTeam, async (id: Schema.Types.ObjectId) => {
    socket.emit("action", {
      type: updated,
      data: {
        [id.toString()]: await getTeam(id)
      }
    });
  });
};
