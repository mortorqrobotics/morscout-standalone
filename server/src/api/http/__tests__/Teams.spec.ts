import { setup, teardown } from "setup/mongoose";
import Teams from "../Teams";
import express from "express";
import request from "supertest";
import { User, Team } from "models";

beforeAll(setup);
afterAll(teardown);

let router: express.Router;

beforeEach(() => {
  router = express.Router();
  jest.spyOn(router, "get");
  Teams(router);
});

test("teams has the same routes", async () => {
  expect(
    (router.get as jest.Mock).mock.calls.map(val => val[0]).sort()
  ).toMatchSnapshot();
});

test("get teams by user works", async () => {
  const user = await User.findOne();
  const req = await request(router).get(`/teams/${user._id}`);
  expect(req.status).toBe(200);
  expect(req.text).toMatchSnapshot();
});

test("get team by id works", async () => {
  const team = await Team.findOne();
  const req = await request(router).get(`/team/${team._id}`);
  expect(req.status).toBe(200);
  expect(JSON.parse(req.text)).toMatchSnapshot({
    _id: expect.any(String),
    location: {
      _id: expect.any(String)
    }
  });
});
