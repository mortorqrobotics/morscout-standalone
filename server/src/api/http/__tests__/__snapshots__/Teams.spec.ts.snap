// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get team by id works 1`] = `
Object {
  "__v": 0,
  "_id": Any<String>,
  "awards": Object {},
  "location": Object {
    "_id": Any<String>,
    "coordinates": Array [
      0,
      0,
    ],
    "type": "Point",
  },
  "name": "MorTorq",
  "number": 1515,
  "regional": "6d7be5a7994e35048b9e5e0e",
  "registeredRegionals": Array [],
}
`;

exports[`get teams by user works 1`] = `"[]"`;

exports[`teams has the same routes 1`] = `
Array [
  "/team/:team",
  "/teams/:season",
]
`;
